app/models/package_manager/multiple_sources_base/provider_map.rb

```diff
@@ -0,0 +1,66 @@
+# frozen_string_literal: true
+
+module PackageManager
+  class MultipleSourcesBase < Base
+    # Given a set of prioritized ProviderInfo objects, determine the most
+    # likely upstream package repository for a package. This uses the
+    # packages' versions' repository_sources field to
+    class ProviderMap
+      # @param prioritized_provider_infos [Array(ProviderInfo)]
+      def initialize(*prioritized_provider_infos)
+        @prioritized_provider_infos = prioritized_provider_infos
+
+        raise "Need at least one provider marked default" if @prioritized_provider_infos.none?(&:default?)
```

Comment:
Should there be some sort of check if there are multiple default providers for one ecosystem?